### PR TITLE
Sodium 0.6 Fabric

### PR DESCRIPTION
**Pull request by IThundxr**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -68,9 +71,11 @@ dependencies {
     // switch to implementation for debugging
     modCompileOnly 'maven.modrinth:starlight:1.1.2+1.20'
 
-    modCompileOnly 'maven.modrinth:sodium:mc1.20.1-0.5.3'
+    modCompileOnly ':sodium-fabric-0.6.0-snapshot+mc1.20.1-local'
     modCompileOnly 'maven.modrinth:iris:1.6.10+1.20.1'
 
+    modCompileOnly 'maven.modrinth:embeddium:kJkE9dfS'
```
Comment (PepperCode1) [inline build.gradle:77]: Use the version number (`0.3.25+mc1.20.1`) instead of the version ID and move this line to be right after the Sodium dependency without additional newlines.

File: src/main/java/com/jozufozu/flywheel/Flywheel.java [hunk 2]
```diff
@@ -39,6 +41,8 @@ public class Flywheel implements ClientModInitializer {
 	public static final Logger LOGGER = LogUtils.getLogger();
 	private static Version version;
 
+
+
```
Comment (PepperCode1) [inline src/main/java/com/jozufozu/flywheel/Flywheel.java]: Unnecessary whitespace.

File: src/main/java/com/jozufozu/flywheel/compat/CompatHelper.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+				Version sodiumVersion = FabricLoader.getInstance()
+						.getModContainer("sodium")
+						.orElseThrow()
+						.getMetadata()
+						.getVersion();
+				IS_SODIUM_0_5 = predicate0_5.test(sodiumVersion);
+				IS_SODIUM_0_6 = predicate0_6.test(sodiumVersion);
+			} catch (Throwable ignored) {}
+		}
+	}
```
Comment (PepperCode1) [inline src/main/java/com/jozufozu/flywheel/compat/CompatHelper.java]: `IS_SODIUM_*` should be final and the logic to compute the values of these fields should be in a static initializer block located right before these fields to prevent ordering issues. `initSodiumVersion` should be removed. Use `VersionPredicate.parse` instead of `VersionPredicateParser.parse` to avoid using implementation classes. Retrieve the Sodium version first, then continue with the logic only if the version was retrieved without an exception.

File: src/main/java/com/jozufozu/flywheel/mixin/sodium/ChunkBuilderMeshingTaskMixin.java [hunk 4]
```diff
@@ -2,22 +2,33 @@
...
-	private BlockEntityRenderer<?> flywheel$redirectGetRenderer(BlockEntityRenderDispatcher dispatcher, BlockEntity blockEntity) {
+	@WrapOperation(
+			method = "execute(Lme/jellysquid/mods/sodium/client/render/chunk/compile/ChunkBuildContext;Lme/jellysquid/mods/sodium/client/util/task/CancellationToken;)Lme/jellysquid/mods/sodium/client/render/chunk/compile/ChunkBuildOutput;",
+			at = @At(
+					value = "INVOKE",
+					target = "Lnet/minecraft/client/renderer/blockentity/BlockEntityRenderDispatcher;getRenderer(Lnet/minecraft/world/level/block/entity/BlockEntity;)Lnet/minecraft/client/renderer/blockentity/BlockEntityRenderer;",
+					remap = true
+			)
+	)
+	private BlockEntityRenderer<?> flywheel$redirectGetRenderer(BlockEntityRenderDispatcher instance, BlockEntity blockEntity, Operation<BlockEntityRenderer<BlockEntity>> original) {
```
Comment (PepperCode1) [inline src/main/java/com/jozufozu/flywheel/mixin/sodium/ChunkBuilderMeshingTaskMixin.java]: Rename the method to `flywheel$wrapGetRenderer`.

File: src/main/resources/fabric.mod.json [hunk 5]
```diff
@@ -31,12 +31,13 @@
...
   "depends": {
-    "fabricloader": ">=0.11.3",
+    "fabricloader": ">=0.15.0",
     "fabric": ">=0.84.0",
     "minecraft": ">=1.20.1",
     "java": ">=17"
   },
   "breaks": {
-    "sodium": "<0.5.0"
+    "sodium": ["<0.5.0", "<0.6 >=0.6.0-beta.2"],
```
Comment (PepperCode1) [inline src/main/resources/fabric.mod.json]: Is this correct? Based on the specification, this means "breaks if Sodium is (<0.5.0 OR (<0.6 AND >=0.6.0-beta.2))", which doesn't make sense because Sodium cannot be older than 0.6 and newer than or equal to 0.6.0-beta.2 at the same time. Also, use `0.6.0` instead of `0.6`.

**Review by PepperCode1 — CHANGES_REQUESTED** [review]

File: build.gradle (hunk 1 above)
Comment (IThundxr) [reply build.gradle:77]: Using the version number will pull the forge version since they have the same version numbers

File: src/main/resources/fabric.mod.json (hunk 5 above)
Comment (PepperCode1) [reply src/main/resources/fabric.mod.json]: Correction: Sodium can be older than 0.6.0 and newer than or equal to 0.6.0-beta.2 at the same time because 0.6.0-beta.2 is a prerelease version; however, this still doesn't make sense because this range would break with all versions newer than or equal to beta 2 instead of older than beta 2 as it is supposed to.

Comment (IThundxr) [reply src/main/resources/fabric.mod.json]: `>0.6.0 <0.6.0-beta.2` should work then i think? if greater than 0.6.0 and less than 0.6.0-beta.2?

Comment (PepperCode1) [reply src/main/resources/fabric.mod.json]: No, because a version cannot be greater than 0.6.0 and less than 0.6.0-beta.2 at the same time.

Comment (PepperCode1) [reply src/main/resources/fabric.mod.json]: I think `["<0.5.0", "~0.6.0 <0.6.0-beta.2"]` will work correctly. Please check that this is actually the case though; try testing this predicate with `0.4.9`, `0.5.11`, `0.6.0-beta.1`, `0.6.0-beta.2`, and `0.6.0` specifically in Minecraft as online tools may not match Fabric Loader's behavior.

File: src/main/java/com/jozufozu/flywheel/compat/CompatHelper.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+					.getModContainer("sodium");
+
+			if (containerOptional.isPresent()) {
+				Version sodiumVersion = containerOptional.get()
+						.getMetadata()
+						.getVersion();
+
+				try {
+					VersionPredicate predicate0_5 = VersionPredicate.parse(">=0.5.0 <0.6.0");
+					VersionPredicate predicate0_6 = VersionPredicate.parse("<0.6 >=0.6.0-beta.2");
```
Comment (PepperCode1) [inline src/main/java/com/jozufozu/flywheel/compat/CompatHelper.java]: These ranges are also incorrect. The first one should be `~0.5.0 >=0.5.0` and the second one should be `>=0.6.0-beta.2`. Again, double check specifically in Minecraft that this is correct.

File: src/main/resources/fabric.mod.json (hunk 5 above)
Comment (IThundxr) [reply src/main/resources/fabric.mod.json]: `VersionPredicate.parse("~0.6.0 <0.6.0-beta.2").test(Version.parse("0.6.0-beta.1"))` returns false when it should be returning true so something else needs to be done for the 0.6.0 breaks

**Comment by PepperCode1** [comment]
Thanks.
